src/main/java/org/gaul/s3proxy/junit/rules/S3ProxyRule.java

```diff
@@ -0,0 +1,188 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy.junit.rules;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.util.Properties;
+
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.gaul.s3proxy.AuthenticationType;
+import org.gaul.s3proxy.S3Proxy;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStoreContext;
+import org.junit.rules.ExternalResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Emulates a S3 endpoint.
+ */
+public final class S3ProxyRule extends ExternalResource {
+    private static final Logger logger = LoggerFactory.getLogger(
+        S3ProxyRule.class);
+
+    private static final String LOCALHOST = "127.0.0.1";
+
+    private final String accessKey;
+    private final String secretKey;
+    private final String endpointFormat;
+    private final File blobStoreLocation;
+    private final S3Proxy s3Proxy;
+
+    public static final class Builder {
+        private AuthenticationType authType = AuthenticationType.NONE;
+        private String accessKey;
+        private String secretKey;
+        private String secretStorePath;
+        private String secretStorePassword;
+        private int port = -1;
+        private boolean ignoreUnknownHeaders;
+        private boolean useHttps;
+        private String blobStoreProvider = "filesystem";
+
+        private Builder() { }
+
+        public Builder withCredentials(AuthenticationType authType,
+                String accessKey, String secretKey) {
+            this.authType = authType;
+            this.accessKey = accessKey;
+            this.secretKey = secretKey;
+            return this;
+        }
+
+        public Builder withCredentials(String accessKey, String secretKey) {
+            return withCredentials(AuthenticationType.AWS_V2_OR_V4, accessKey,
+                secretKey);
+        }
+
+        public Builder withSecretStore(String path, String password) {
+            secretStorePath = path;
+            secretStorePassword = password;
+            return this;
+        }
+
+        public Builder withPort(int port) {
+            this.port = port;
+            return this;
+        }
+
+        public Builder withBlobStoreProvider(String blobStoreProvider) {
+            this.blobStoreProvider = blobStoreProvider;
+            return this;
+        }
+
+        public Builder withHttps() {
+            useHttps = true;
+            return this;
+        }
+
+        public Builder ignoreUnknownHeaders() {
+            ignoreUnknownHeaders = true;
+            return this;
+        }
+
+        public S3ProxyRule build() {
+            return new S3ProxyRule(this);
```

Comment:
I still question the need for a second builder with slightly different signatures.  Instead can you just add a method `S3ProxyRule.newRule(S3Proxy.Builder)` which already supports all these flags and more?

Reply:
I guess this comes down to our use cases for the rule so far, most of our tests don't want/meed to know any of the details of S3Proxy and want a bare minimum S3 endpoint that they can interact with. Currently all they have to do in order to achieve this is put the following in a test:

    public @Rule S3ProxyRule s3Proxy = S3ProxyRule
        .builder()
        .withCredentials(AWS_ACCESS_KEY, AWS_SECRET_KEY)
        .build();

If they had to go via the S3ProxyBuilder they would need to potentially do all the stuff that the S3ProxyRule constructor is doing (see lines 109-151) which kind of defeats the whole point of the rule.